File: dotnet/ActorId.cs
namespace Epoch;

public readonly struct ActorIdParts
{
    public ActorIdParts(int region, int server, int processType, int processIndex, int actorIndex)
    {
        Region = region;
        Server = server;
        ProcessType = processType;
        ProcessIndex = processIndex;
        ActorIndex = actorIndex;
    }

    public int Region { get; }
    public int Server { get; }
    public int ProcessType { get; }
    public int ProcessIndex { get; }
    public int ActorIndex { get; }
}

public interface IActorIdCodec
{
    long Encode(ActorIdParts parts);
    ActorIdParts Decode(long value);
    string Name { get; }
}

public sealed class DefaultActorIdCodec : IActorIdCodec
{
    private const int RegionBits = 10;
    private const int ServerBits = 12;
    private const int ProcessTypeBits = 6;
    private const int ProcessIndexBits = 10;
    private const int ActorIndexBits = 26;

    private const ulong RegionMax = (1UL << RegionBits) - 1;
    private const ulong ServerMax = (1UL << ServerBits) - 1;
    private const ulong ProcessTypeMax = (1UL << ProcessTypeBits) - 1;
    private const ulong ProcessIndexMax = (1UL << ProcessIndexBits) - 1;
    private const ulong ActorIndexMax = (1UL << ActorIndexBits) - 1;

    private const int ActorIndexShift = 0;
    private const int ProcessIndexShift = ActorIndexShift + ActorIndexBits;
    private const int ProcessTypeShift = ProcessIndexShift + ProcessIndexBits;
    private const int ServerShift = ProcessTypeShift + ProcessTypeBits;
    private const int RegionShift = ServerShift + ServerBits;

    public static readonly DefaultActorIdCodec Instance = new();

    public long Encode(ActorIdParts parts)
    {
        if (parts.Region < 0 || (ulong)parts.Region > RegionMax ||
            parts.Server < 0 || (ulong)parts.Server > ServerMax ||
            parts.ProcessType < 0 || (ulong)parts.ProcessType > ProcessTypeMax ||
            parts.ProcessIndex < 0 || (ulong)parts.ProcessIndex > ProcessIndexMax ||
            parts.ActorIndex < 0 || (ulong)parts.ActorIndex > ActorIndexMax)
        {
            throw new ArgumentOutOfRangeException(nameof(parts), "ActorId parts out of range");
        }

        ulong value = ((ulong)parts.Region << RegionShift)
            | ((ulong)parts.Server << ServerShift)
            | ((ulong)parts.ProcessType << ProcessTypeShift)
            | ((ulong)parts.ProcessIndex << ProcessIndexShift)
            | ((ulong)parts.ActorIndex << ActorIndexShift);

        return unchecked((long)value);
    }

    public ActorIdParts Decode(long value)
    {
        ulong raw = unchecked((ulong)value);
        int region = (int)((raw >> RegionShift) & RegionMax);
        int server = (int)((raw >> ServerShift) & ServerMax);
        int processType = (int)((raw >> ProcessTypeShift) & ProcessTypeMax);
        int processIndex = (int)((raw >> ProcessIndexShift) & ProcessIndexMax);
        int actorIndex = (int)((raw >> ActorIndexShift) & ActorIndexMax);
        return new ActorIdParts(region, server, processType, processIndex, actorIndex);
    }

    public string Name => "default";
}

public static class ActorId
{
    public static IActorIdCodec Default => DefaultActorIdCodec.Instance;

    public static long Encode(ActorIdParts parts) => Default.Encode(parts);

    public static ActorIdParts Decode(long value) => Default.Decode(value);
}

File: cpp/src/actor_id.cpp
#include "epoch/actor_id.h"

#include <stdexcept>

namespace epoch {

namespace {
constexpr std::uint32_t kRegionBits = 10;
constexpr std::uint32_t kServerBits = 12;
constexpr std::uint32_t kProcessTypeBits = 6;
constexpr std::uint32_t kProcessIndexBits = 10;
constexpr std::uint32_t kActorIndexBits = 26;

constexpr std::uint64_t kRegionMax = (1ULL << kRegionBits) - 1;
constexpr std::uint64_t kServerMax = (1ULL << kServerBits) - 1;
constexpr std::uint64_t kProcessTypeMax = (1ULL << kProcessTypeBits) - 1;
constexpr std::uint64_t kProcessIndexMax = (1ULL << kProcessIndexBits) - 1;
constexpr std::uint64_t kActorIndexMax = (1ULL << kActorIndexBits) - 1;

constexpr std::uint32_t kActorIndexShift = 0;
constexpr std::uint32_t kProcessIndexShift = kActorIndexShift + kActorIndexBits;
constexpr std::uint32_t kProcessTypeShift = kProcessIndexShift + kProcessIndexBits;
constexpr std::uint32_t kServerShift = kProcessTypeShift + kProcessTypeBits;
constexpr std::uint32_t kRegionShift = kServerShift + kServerBits;
} // namespace

std::uint64_t DefaultActorIdCodec::encode(const ActorIdParts &parts) const
{
    if (parts.region > kRegionMax || parts.server > kServerMax ||
        parts.process_type > kProcessTypeMax || parts.process_index > kProcessIndexMax ||
        parts.actor_index > kActorIndexMax)
    {
        throw std::out_of_range("ActorIdParts out of range");
    }

    return (static_cast<std::uint64_t>(parts.region) << kRegionShift) |
           (static_cast<std::uint64_t>(parts.server) << kServerShift) |
           (static_cast<std::uint64_t>(parts.process_type) << kProcessTypeShift) |
           (static_cast<std::uint64_t>(parts.process_index) << kProcessIndexShift) |
           (static_cast<std::uint64_t>(parts.actor_index) << kActorIndexShift);
}

ActorIdParts DefaultActorIdCodec::decode(std::uint64_t value) const
{
    ActorIdParts parts{};
    parts.region = static_cast<std::uint16_t>((value >> kRegionShift) & kRegionMax);
    parts.server = static_cast<std::uint16_t>((value >> kServerShift) & kServerMax);
    parts.process_type = static_cast<std::uint8_t>((value >> kProcessTypeShift) & kProcessTypeMax);
    parts.process_index = static_cast<std::uint16_t>((value >> kProcessIndexShift) & kProcessIndexMax);
    parts.actor_index = static_cast<std::uint32_t>((value >> kActorIndexShift) & kActorIndexMax);
    return parts;
}

const char *DefaultActorIdCodec::name() const
{
    return "default";
}

const ActorIdCodec &default_actor_id_codec()
{
    static DefaultActorIdCodec codec;
    return codec;
}

std::uint64_t encode_actor_id(const ActorIdParts &parts, const ActorIdCodec &codec)
{
    return codec.encode(parts);
}

ActorIdParts decode_actor_id(std::uint64_t value, const ActorIdCodec &codec)
{
    return codec.decode(value);
}

} // namespace epoch
